File: algorithm/go/firstround/77.go
package firstround

func combine(n, k int) [][]int {
	if n < 1 || k < 1 {
		return nil
	}

	res := make([][]int, 0)
	track := make([]int, 0)
	backtrack(&res, track, n, k, 1)
	return res
}

func backtrack(res *[][]int, track []int, n, k, idx int) {
	if len(track) == k {
		tmp := make([]int, k)
		copy(tmp, track)
		*res = append(*res, tmp)
		return
	}

	for i := idx; i <= n; i++ {
		track = append(track, i)
		backtrack(res, track, n, k, i+1)
		track = track[:len(track)-1]
	}
}

File: algorithm/cpp/firstround/77.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> combine(int n, int k);

private:
    void recur(int start);
    vector<vector<int>> res;
    vector<int> cur;
    int n;
    int k;
};

vector<vector<int>> Solution::combine(int n, int k) {
    if(k>n) return {};
    else if(n==k)
    {
        vector<int> tmp;
        for(int i=1; i<=n; i++)
        {
            tmp.emplace_back(i);
        }
        return {tmp};
    }

    this->n = n;
    this->k = k;
    recur(1);
    return res;
}

void Solution::recur(int start)
{
    if(cur.size()==k)
    {
        res.emplace_back(cur);
        return;
    }

    int end = n-(k-cur.size())+1;
    for(int i=start; i<=end; i++)
    {
        cur.emplace_back(i);
        recur(i+1);
        cur.pop_back();
    }
    return ;
}
int main()
{
    int n = 4;
    int k = 2;
    Solution obj;
    vector<vector<int>> result = obj.combine(n, k);
    return 0;
}